File: binary_search_step.py
import csv
import os

def load_dataset(filename):
    with open(filename, newline='') as file:
        reader = csv.reader(file)
        return [(int(row[0]), row[1]) for row in reader]

def binary_search_step(dataset, target, output_filename):
    left = 0
    right = len(dataset) - 1

    with open(output_filename, "w") as f:
        while left <= right:
            mid = (left + right) // 2
            number, text = dataset[mid]

            f.write(f"{mid + 1}: {number}/{text}\n")

            if number == target:
                return
            elif number < target:
                left = mid + 1
            else:
                right = mid - 1

        f.write("-1\n")

# Get the directory where this script is located
script_dir = os.path.dirname(os.path.abspath(__file__))

csv_files = [f for f in os.listdir(script_dir) if f.lower().endswith(".csv")]

if not csv_files:
    print("No CSV files found in the current directory.")
    exit()

# Display options to user
print("Choose a dataset:")
for idx, fname in enumerate(csv_files, start=1):
    print(f"{idx}. {fname}")

try:
    choice = int(input(f"Enter choice (1-{len(csv_files)}): "))
    if choice < 1 or choice > len(csv_files):
        raise ValueError
except ValueError:
    print("Invalid choice.")
    exit()

filename = csv_files[choice - 1]
full_path = os.path.join(script_dir, filename)

# Ask for target
try:
    target = int(input("Enter target value to search: "))
except ValueError:
    print("Invalid input. Please enter an integer.")
    exit()

output_file = f"binary_search_step_{target}.txt"

data = load_dataset(full_path)

binary_search_step(data, target, output_file)

print(f"Search steps saved to {output_file}")

File: binary_search_step.java
import java.io.*;
import java.util.*;

public class binary_search_step {

    static class DataRow {
        int number;
        String word;

        DataRow(int number, String word) {
            this.number = number;
            this.word = word;
        }
    }

    public static void main(String[] args) throws IOException {
        Scanner scanner = new Scanner(System.in);

        // Get current working directory
        String currentDir = System.getProperty("user.dir");

        // List all CSV files in the directory
        File dir = new File(currentDir);
        File[] files = dir.listFiles((d, name) -> name.toLowerCase().endsWith(".csv"));

        if (files == null || files.length == 0) {
            System.out.println("No CSV files found in the current directory.");
            return;
        }

        // Display dataset options
        System.out.println("Choose a dataset:");
        for (int i = 0; i < files.length; i++) {
            System.out.println((i + 1) + ". " + files[i].getName());
        }

        System.out.print("Enter your choice (1-" + files.length + "): ");
        int choice = scanner.nextInt();
        scanner.nextLine(); 

        if (choice < 1 || choice > files.length) {
            System.out.println("Invalid choice.");
            return;
        }

        String filename = files[choice - 1].getAbsolutePath();

        // Prompt user to enter the target integer
        System.out.print("Enter target integer to search: ");
        int target = scanner.nextInt();

        String outputFile = "binary_search_step_" + target + ".txt";

        List<DataRow> dataset = loadCSV(filename);
        binarySearchStep(dataset, target, outputFile);

        System.out.println("Search steps saved to " + outputFile);
    }

    public static List<DataRow> loadCSV(String filename) throws IOException {
        List<DataRow> list = new ArrayList<>();
        BufferedReader br = new BufferedReader(new FileReader(filename));
        String line;
        boolean skipFirst = true; // Skip header if present

        while ((line = br.readLine()) != null) {
            String[] parts = line.split(",");
            try {
                int number = Integer.parseInt(parts[0]);
                String word = parts[1];
                list.add(new DataRow(number, word));
            } catch (NumberFormatException e) {
                if (skipFirst) {
                    skipFirst = false;
                    continue;
                }
            }
        }
        br.close();
        return list;
    }

    public static void binarySearchStep(List<DataRow> data, int target, String outputFile) throws IOException {
        int left = 0;
        int right = data.size() - 1;

        BufferedWriter writer = new BufferedWriter(new FileWriter(outputFile));

        while (left <= right) {
            int mid = (left + right) / 2;
            DataRow row = data.get(mid);
            writer.write((mid + 1) + ": " + row.number + "/" + row.word + "\n");

            if (row.number == target) {
                writer.close();
                return;
            } else if (row.number < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        // Target not found
        writer.write("-1\n");
        writer.close();
    }
}
